### PR TITLE
improve compile errors for invalid ptr-to-ptr casts with trait objects

### LINKED ISSUE #130030: Confusing diagnostic for invalid casts of pointers to trait objects
### Code

```rust
fn main() {
    let ptr: *const dyn Fn() = &|| ();
    _ = ptr as *const dyn Fn(u8);
}
```

### Current output

```
error[E0308]: mismatched types
 --> src/main.rs:3:9
  |
3 |     _ = ptr as *const dyn Fn(u8);
  |         ^^^^^^^^^^^^^^^^^^^^^^^^ expected `()`, found `(u8,)`
  |
  = note: expected trait object `dyn Fn()`
             found trait object `dyn Fn(u8)`
```

### Desired output

```
error[E0308]: invalid cast from `*const dyn Fn()` to `*const dyn Fn(u8)`
 --> src/main.rs:3:9
  |
3 |     _ = ptr as *const dyn Fn(u8);
  |         ^^^^^^^^^^^^^^^^^^^^^^^^ expected `()`, found `(u8,)`
  |
  = note: `Fn()` and `Fn(u8)` might have different vtables, leading to unsoundness
  = note: if you are absolutely sure you are using the resulting value safely, use `transmute`
```

### Rationale and extra context

In #120248 I made checks for casting pointers to trait objects more strict: now they need to have the same generic arguments (this is due to the fact that for the cast to be safe, we need vtables to match).

The check is however made in a bit of a crude way, which results in this "mismatched types" diagnostic.

We should modify it so that the intent of the error is clear.

### Rust Version

Reproduces on both stable 1.81.0 and 1.83.0-nightly (2024-09-05 9c01301c52df5d2d7b6f).

### PR DESCRIPTION
**Pull request by lukas-code**
This is a follow-up to https://github.com/rust-lang/rust/pull/120248 to improve some of its error messages.

1. Make the borrowcheck error for "type annotation requires that x must outlive y" actually point at the type annotation, i.e. the type `T` in a `x as T` cast. This makes the error more consistent with other errors caused by type annotation in other places, such as
```text
error: lifetime may not live long enough
 --> src/lib.rs:4:12
  |
3 | fn bar(a: &i32) {
  |           - let's call the lifetime of this reference `'1`
4 |     let b: &'static i32 = a;
  |            ^^^^^^^^^^^^ type annotation requires that `'1` must outlive `'static`
```

2. Don't say "cast" when we actually mean "coercion" and give borrowcheck errors from actual casts (which is currently just the check added in https://github.com/rust-lang/rust/pull/120248) a higher priority than ones from coercions. This can improve the errors for ptr-to-ptr cast between trait objects because they are are lowered as an upcast "unsizing" coercion if possible (which may be the identity upcast) followed by the actual cast.

3. Bring back the old "casting X as Y is invalid" message for type mismatch in the principals and reword the "vtable kinds may not match" message to more accurately describe the pointer metadata and not refer to "vtables" if the metadata is unknown.

fixes https://github.com/rust-lang/rust/issues/130030

r? @WaffleLapkin but feel free to reassign

### CODE REVIEW

File: compiler/rustc_hir_typeck/src/cast.rs [hunk 1]
```diff
@@ -306,10 +310,26 @@ impl<'a, 'tcx> CastCheck<'tcx> {
...
+                    }
+                    (PointerKind::OfParam(_) | PointerKind::OfAlias(_), _)
+                    | (_, PointerKind::OfParam(_) | PointerKind::OfAlias(_)) => {
+                        err.note("the pointers may have different metadata");
+                    }
+                    (PointerKind::VTable(_), PointerKind::Length)
+                    | (PointerKind::Length, PointerKind::VTable(_)) => {
+                        err.note("the pointers have different metadata");
+                    }
+                    _ => span_bug!(self.span, "unexpected cast error: {e:?}"),
```
Comment (WaffleLapkin) [inline compiler/rustc_hir_typeck/src/cast.rs:329]: Can you make this match exhaustive?

File: compiler/rustc_middle/src/mir/query.rs [hunk 2]
```diff
@@ -233,8 +233,9 @@ pub enum ConstraintCategory<'tcx> {
     UseAsConst,
     UseAsStatic,
     TypeAnnotation,
-    Cast {
-        /// Whether this is an unsizing cast and if yes, this contains the target type.
+    Cast,
```
Comment (WaffleLapkin) [inline compiler/rustc_middle/src/mir/query.rs:236]: Why does a cast not have `unsize_to`? Casts can be unsizing, for example: `&() as &dyn std::any::Any`.

File: compiler/rustc_middle/src/thir/visit.rs [hunk 3]
```diff
@@ -129,7 +129,7 @@ pub fn walk_expr<'thir, 'tcx: 'thir, V: Visitor<'thir, 'tcx>>(
                 visitor.visit_expr(&visitor.thir()[base.base]);
             }
         }
-        PlaceTypeAscription { source, user_ty: _ } | ValueTypeAscription { source, user_ty: _ } => {
+        PlaceTypeAscription { source, .. } | ValueTypeAscription { source, .. } => {
```
Comment (WaffleLapkin) [inline compiler/rustc_middle/src/thir/visit.rs:132]: Please undo this change. (note how other things in this match always include all fields)

File: compiler/rustc_mir_build/src/build/expr/as_place.rs [hunk 4]
```diff
@@ -499,22 +500,23 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
...
-                            span: source_info.span,
+                            span: user_ty_span,
                             user_ty: user_ty.clone(),
                             inferred_ty: expr.ty,
                         });
                     this.cfg.push(
                         block,
                         Statement {
-                            source_info,
+                            source_info: ty_source_info,
```
Comment (WaffleLapkin) [inline compiler/rustc_mir_build/src/build/expr/as_place.rs:519]: this feels weird, I don't have much context, but are you sure this is correct? shouldn't `source_info` for a statement cover the whole statement?

File: compiler/rustc_mir_build/src/thir/print.rs [hunk 5]
```diff
@@ -454,14 +454,14 @@ impl<'a, 'tcx> ThirPrinter<'a, 'tcx> {
...
-            PlaceTypeAscription { source, user_ty } => {
+            PlaceTypeAscription { source, user_ty, .. } => {
                 print_indented!(self, "PlaceTypeAscription {", depth_lvl);
                 print_indented!(self, format!("user_ty: {:?}", user_ty), depth_lvl + 1);
                 print_indented!(self, "source:", depth_lvl + 1);
                 self.print_expr(*source, depth_lvl + 2);
                 print_indented!(self, "}", depth_lvl);
             }
-            ValueTypeAscription { source, user_ty } => {
+            ValueTypeAscription { source, user_ty, .. } => {
```
Comment (WaffleLapkin) [inline compiler/rustc_mir_build/src/thir/print.rs:464]: `Call` seems to print span too. You should probably do this here as well.

File: tests/ui/lifetimes/issue-90600-expected-return-static-indirect.stderr [hunk 6]
```diff
@@ -7,7 +7,7 @@ LL |     let refcell = RefCell::new(&mut foo);
    |                                ^^^^^^^^ borrowed value does not live long enough
 LL |
 LL |     let read = &refcell as &RefCell<dyn Read>;
-   |                -------- cast requires that `foo` is borrowed for `'static`
+   |                -------- coercion requires that `foo` is borrowed for `'static`
```
Comment (WaffleLapkin) [inline tests/ui/lifetimes/issue-90600-expected-return-static-indirect.stderr:10]: this and some others bellow sure look like a cast?

File: tests/ui/traits/trait-object-lifetime-default-note.rs [hunk 7]
```diff
@@ -8,7 +8,7 @@ fn main() {
     //~| NOTE borrowed value does not live long enough
     //~| NOTE due to object lifetime defaults, `Box<dyn A>` actually means `Box<(dyn A + 'static)>`
     require_box(Box::new(r));
-    //~^ NOTE cast requires that `local` is borrowed for `'static`
+    //~^ NOTE requires that `local` is borrowed for `'static`
```
Comment (WaffleLapkin) [inline tests/ui/traits/trait-object-lifetime-default-note.rs:11]: you should include "coercion"

File: compiler/rustc_hir_typeck/src/cast.rs [hunk 8]
```diff
@@ -865,7 +892,15 @@ impl<'a, 'tcx> CastCheck<'tcx> {
                         ));
 
                         // `dyn Src = dyn Dst`, this checks for matching traits/generics
-                        fcx.demand_eqtype(self.span, src_obj, dst_obj);
+                        let cause = fcx.misc(self.span);
```
Comment (compiler-errors) [inline compiler/rustc_hir_typeck/src/cast.rs:895]: Leave a comment saying that this is just `demand_eqtype` but inlined so it has a better error

File: compiler/rustc_mir_build/src/build/expr/as_place.rs (hunk 4 above)
Comment (lukas-code) [reply compiler/rustc_mir_build/src/build/expr/as_place.rs:519]: This matches what we do `AscribeUserType` in other places like the type of [let statements](https://github.com/rust-lang/rust/blob/7c7372b6a1b9a8d6b787a3e5a10e4f8325b1fdfc/compiler/rustc_mir_build/src/build/matches/mod.rs#L638) or [patterns](https://github.com/rust-lang/rust/blob/7c7372b6a1b9a8d6b787a3e5a10e4f8325b1fdfc/compiler/rustc_mir_build/src/build/matches/mod.rs#L2559).

Comment (WaffleLapkin) [reply compiler/rustc_mir_build/src/build/expr/as_place.rs:519]: Okay 👍🏻

File: compiler/rustc_middle/src/mir/query.rs (hunk 2 above)
Comment (lukas-code) [reply compiler/rustc_middle/src/mir/query.rs:236]: The intent was to have `Cast` represent only the non-coercion casts from the table [here](https://doc.rust-lang.org/nightly/reference/expressions/operator-expr.html#type-cast-expressions)[^1] and use `Coercion` for any coercion whether they are caused by a cast or not. But seeing that MIR also just represents coercions as special kind of cast and not the other way around, maybe we should just add a field `is_coercion` to the `Cast` variant and not add a new variant.

[^1]: side note: function item to function pointer and closure to function pointer should not be listed there as they are coercions.

File: tests/ui/lifetimes/issue-90600-expected-return-static-indirect.stderr (hunk 6 above)
Comment (lukas-code) [reply tests/ui/lifetimes/issue-90600-expected-return-static-indirect.stderr:10]: This is a side effect from how explicit coercion casts and implicit coercions are represented in the same way in MIR. I will try to recover the wording here.

File: compiler/rustc_middle/src/mir/query.rs (hunk 2 above)
Comment (WaffleLapkin) [reply compiler/rustc_middle/src/mir/query.rs:236]: > maybe we should just add a field `is_coercion` to the `Cast` variant and not add a new variant.

This seems appropriate. I think it's nice to mention what the user actually written in syntax. But trying to explain that `as` casts are a super set of coercions is probably confusing.

File: tests/ui/lifetimes/issue-90600-expected-return-static-indirect.stderr (hunk 6 above)
Comment (lukas-code) [reply tests/ui/lifetimes/issue-90600-expected-return-static-indirect.stderr:10]: So this turned out to be non-trivial, because there is currently no way to distinguish
```rust
let x: T = y;
```
from
```rust
let x = y as T
```
by looking at the MIR alone (without relying on the exact order of MIR statements).

So our options here are
* Leave the diagnostic as is, calling some coercion casts just "coercions" (technically correct).
* Revert the wording and call all coercions "casts" (technically wrong).
* Add some additional diagnostic-only information into the MIR to allow us to distinguish them and output the optimal diagnostic (which is really just one word). This is currently unprecedented as far as I'm aware.

I'm unsure what the best option here is and would appreciate your input :)

Comment (WaffleLapkin) [reply tests/ui/lifetimes/issue-90600-expected-return-static-indirect.stderr:10]: I would prefer the third option, i.e. to add some diagnostic only info. That being said I think you should check with people who work on MIR more (miri or miropt zulip streams maybe?).

The first option (using "coercion" when the user wrote a cast) could be OK. If you want we can leave the third option as a follow up (please try to do it though 😄).

Comment (lukas-code) [reply tests/ui/lifetimes/issue-90600-expected-return-static-indirect.stderr:10]: After checking again, I found that we do have one instance of diagnostic information in MIR here, so this feels less unprecedented now: https://github.com/rust-lang/rust/blob/8c2c9a9ef527edb453bcee638d34beb0e7b4616b/compiler/rustc_middle/src/mir/syntax.rs#L735-L736

**Comment by lukas-code** [comment]
Ah shit. Well I guess that's one way to ask for feedback about diagnostic info in MIR. :smiling_face_with_tear:

The relevant commit is 59696de25b18fab9127eb089a43673050ad0cf06.

File: tests/ui/dyn-star/dyn-to-rigid.stderr [hunk 9]
```diff
@@ -1,9 +1,9 @@
-error[E0606]: casting `(dyn* Tr + 'static)` as `usize` is invalid
+error[E0605]: non-primitive cast: `(dyn* Tr + 'static)` as `usize`
   --> $DIR/dyn-to-rigid.rs:7:5
    |
 LL |     x as usize
-   |     ^^^^^^^^^^
+   |     ^^^^^^^^^^ an `as` expression can only be used to convert between primitive types or to coerce to a specific trait object
```
Comment (lukas-code) [inline tests/ui/dyn-star/dyn-to-rigid.stderr:5]: This diagnostic is terrible, but that also applies to all other places that it is emitted and I'd rather postpone cleaning it up, because this PR is already getting way too large.

Comment (saethlin) [reply tests/ui/dyn-star/dyn-to-rigid.stderr:5]: I support that.

File: compiler/rustc_middle/src/mir/syntax.rs [hunk 10]
```diff
@@ -1421,6 +1419,15 @@ pub enum CastKind {
     Transmute,
 }
 
+/// Represents how a [`CastKind::PointerCoercion`] was constructed, used for diagnostics.
```
Comment (RalfJung) [inline compiler/rustc_middle/src/mir/syntax.rs:1422]:
```suggestion
/// Represents how a [`CastKind::PointerCoercion`] was constructed.
/// Used only for diagnostics.
```

File: compiler/rustc_borrowck/src/type_check/mod.rs [hunk 11]
```diff
@@ -1977,8 +1977,9 @@ impl<'a, 'tcx> TypeChecker<'a, 'tcx> {
             Rvalue::Cast(cast_kind, op, ty) => {
                 self.check_operand(op, location);
 
-                match cast_kind {
-                    CastKind::PointerCoercion(PointerCoercion::ReifyFnPointer) => {
+                match *cast_kind {
+                    CastKind::PointerCoercion(PointerCoercion::ReifyFnPointer, coercion_source) => {
+                        let is_implicit_coercion = coercion_source == CoercionSource::Implicit;
```
Comment (WaffleLapkin) [inline compiler/rustc_borrowck/src/type_check/mod.rs:1982]: nitpick: Maybe just have a `CoercionSource` field in `ConstraintCategory::Cast` too, instead of a `bool` and this conversion?

File: compiler/rustc_middle/src/mir/syntax.rs [hunk 12]
```diff
@@ -1403,9 +1404,7 @@ pub enum CastKind {
     /// * [`PointerCoercion::MutToConstPointer`]
     ///
     /// Both are runtime nops, so should be [`CastKind::PtrToPtr`] instead in runtime MIR.
-    PointerCoercion(PointerCoercion),
-    /// Cast into a dyn* object.
-    DynStar,
+    PointerCoercion(PointerCoercion, CoercionSource),
```
Comment (WaffleLapkin) [inline compiler/rustc_middle/src/mir/syntax.rs:1407]: I'm not well versed in `dyn*` object, does grouping them with pointers actually make sense? cc @compiler-errors

File: tests/ui/kindck/kindck-impl-type-params.stderr [hunk 13]
```diff
@@ -112,13 +112,13 @@ LL |     struct Foo; // does not impl Copy
...
 error: lifetime may not live long enough
-  --> $DIR/kindck-impl-type-params.rs:30:13
+  --> $DIR/kindck-impl-type-params.rs:30:19
    |
 LL | fn foo<'a>() {
    |        -- lifetime `'a` defined here
 LL |     let t: S<&'a isize> = S(marker::PhantomData);
 LL |     let a = &t as &dyn Gettable<&'a isize>;
-   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ type annotation requires that `'a` must outlive `'static`
+   |                   ^^^^^^^^^^^^^^^^^^^^^^^^ type annotation requires that `'a` must outlive `'static`
```
Comment (WaffleLapkin) [inline tests/ui/kindck/kindck-impl-type-params.stderr:121]: question: why does this say "type annotation" instead of "cast" like most others?

**Review by WaffleLapkin — APPROVED** [review]
I like this a lot!

r=me with or without nits.

File: compiler/rustc_borrowck/src/type_check/mod.rs (hunk 11 above)
Comment (lukas-code) [reply compiler/rustc_borrowck/src/type_check/mod.rs:1982]: I considered this, but decided against it, because the `Ord` of `ConstraintCategory` is relevant for the diagnostic output and I didn't want this property to spread all over the compiler.

File: tests/ui/kindck/kindck-impl-type-params.stderr (hunk 13 above)
Comment (lukas-code) [reply tests/ui/kindck/kindck-impl-type-params.stderr:121]: Because borrow errors from type annotations are given special priority over errors from other sources (including casts) here: https://github.com/rust-lang/rust/blob/1a5a2240bc1b8cf0bcce7acb946c78d6493a4fd3/compiler/rustc_borrowck/src/region_infer/mod.rs#L2105
Changing this code at all makes many other diagnostics worse, so I'd prefer to leave it alone for now.

Comment (lukas-code) [reply tests/ui/kindck/kindck-impl-type-params.stderr:121]: Also, this special casing can cause some diagnostic output that is just plainly wrong, so it should definitely be changed in the future.

```rust
trait Trait<'a> {}

fn require_static(_: *const dyn Trait<'static>) {}

fn not_static<'a>(a: *const dyn Trait<'a>) {
    require_static(a as *const dyn Trait<'a>);
}
```

```
error: lifetime may not live long enough
 --> src/lib.rs:6:20
  |
5 | fn not_static<'a>(a: *const dyn Trait<'a>) {
  |               -- lifetime `'a` defined here
6 |     require_static(a as *const dyn Trait<'a>);
  |                    ^^^^^^^^^^^^^^^^^^^^^^^^^ type annotation requires that `'a` must outlive `'static`
```

**Comment by lukas-code** [comment]
Rebased with no further changes. Per https://github.com/rust-lang/rust/pull/130234#pullrequestreview-2317761695, I'll just go ahead and

@bors r=WaffleLapkin

Any further improvements can be landed separately.
